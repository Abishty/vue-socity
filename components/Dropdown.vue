<template>
  <div id="dropdown" class="dropdown row">
    <div
      v-for="city in cities"
      :key="city.key"
      class="col-md-1 text-center city-name city-choose"
    >
      <div
        :class="city.key === $route.params.city ? 'current-city' : ''"
        @click="handleCityClick(city.key)"
      >
        <img :src="city.img" :class="`city-img ${city.key}`" />
        <p>{{ city.name }}</p>
      </div>
      <div v-if="city.key === $route.params.city">
        <p class="current-text">Current City</p>
      </div>
    </div>
  </div>
</template>

<script>
import constants from '~/constants'
export default {
  data() {
    return {
      cities: constants.CITIES
    }
  },
  mounted() {
    window.addEventListener('click', this.handleOutsideClick)
  },
  destroyed() {
    window.removeEventListener('click', this.handleOutsideClick)
  },
  methods: {
    handleOutsideClick(e) {
      // ignore if arrow is clicked
      if (e.target.closest('#arrow')) {
        return
      }

      if (!e.target.closest('#dropdown')) {
        this.$emit('closeDropdown')
      }
    },
    handleCityClick(city) {
      this.$emit('changeCity', city)
    }
  }
}
</script>

<style scoped>
.dropdown {
  position: fixed;
  width: 88%;
  border: 1px solid black;
  background-color: white;
  z-index: 1;
  top: 70px;
  left: 75px;
  border-radius: 5px;
}
.city-name {
  font-size: 14px;
  margin-top: 15px;
  cursor: pointer;
}
.city-img {
  box-sizing: border-box;
  height: 30px;
  padding-top: 4px;
}
.city-choose {
  box-sizing: border-box;
  width: 9.09%;
  padding: 1% 1% 3%;
  height: 98px;
}
.srinagar {
  width: 55px;
  height: auto;
  padding-top: 12px;
}
.current-city {
  border: 0.75px solid rgb(147, 137, 143);
  box-sizing: border-box;
  height: 55px;
}
.current-text {
  font-size: 12px;
}
a {
  text-decoration: none;
  color: black;
}
</style>
