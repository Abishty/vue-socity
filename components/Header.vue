<template>
  <div>
    <div ref="headerComponent" class="row header-container">
      <div class="col-2 ham-menu" style="text-align: center;">
        <font-awesome-icon :icon="['fas', 'bars']" />
      </div>
      <div class="col-lg-9 col-8 left-side">
        <div @click="$router.back()" class="back-btn">
          <img src="../assets/arrow.svg" />
        </div>
        <div class="logo">
          <nuxt-link to="/">
            <img src="../assets/so_logo.svg" />
          </nuxt-link>
        </div>
        <div class="city-names">{{ cityName }}</div>
        <div id="arrow" @click="toggleDropdown">
          <img src="../assets/arrow.svg" />
        </div>
      </div>
      <div class="col-lg-3 col-2 right-side">
        <div class="write-a-post">Write A Post</div>
        <div class="search">
          <img src="../assets/search.svg" />
        </div>
      </div>
    </div>
    <transition name="fade">
      <Dropdown
        v-if="showDropdown"
        v-on:closeDropdown="toggleDropdown"
        v-on:changeCity="changeCity"
      />
    </transition>
  </div>
</template>
<script>
import Dropdown from './Dropdown'
import constants from '~/constants'

export default {
  components: {
    Dropdown
  },
  data() {
    return {
      city: this.$route.params.city,
      showDropdown: false
    }
  },
  computed: {
    cityName() {
      const city = constants.CITIES.find((c) => c.key === this.city)
      return city ? city.name : 'Delhi'
    }
  },
  watch: {
    $route: {
      handler() {
        this.city = this.$route.params.city
      }
    }
  },
  mounted() {
    this.initialHeaderOffset = this.$refs.headerComponent.offsetTop

    window.addEventListener('scroll', this.handleScroll)

    this.handleScroll()
  },
  methods: {
    toggleDropdown() {
      this.showDropdown = !this.showDropdown
    },
    changeCity(city) {
      this.toggleDropdown()
      this.$router.push(`/${city}`)
    },
    handleScroll() {
      if (window.pageYOffset <= this.initialHeaderOffset) {
        this.$refs.headerComponent.classList.remove('sticky')
        this.$emit('removePaddingTop')
      }
      if (window.pageYOffset > this.$refs.headerComponent.offsetTop) {
        this.$refs.headerComponent.classList.add('sticky')
        this.$emit('addPaddingTop')
      }
    }
  }
}
</script>

<style scoped>
.ham-menu {
  display: none;
}
.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s;
}
.fade-enter, .fade-leave-to /* .fade-leave-active below version 2.1.8 */ {
  opacity: 0;
}
.left-side {
  display: flex;
  align-self: center;
}
.right-side {
  display: flex;
  justify-content: space-evenly;
  align-self: center;
}
.header-container {
  width: 100%;
  z-index: 1;
  background-color: white;
  height: 70px;
  margin-right: 0px;
  margin-left: 0px;
  box-shadow: 0px 0px 3px 0 #000000;
}
.back-btn {
  cursor: pointer;
  width: 20px;
  height: 14px;
  position: relative;
  top: 10px;
}
.back-btn img {
  transform: rotate(180deg);
  height: 100%;
}
.city-names {
  position: relative;
  left: 32px;
  align-self: center;
  font-size: 14px;
}
#arrow {
  cursor: pointer;
  position: relative;
  left: 36px;
  align-self: center;
}
#arrow img {
  width: 12px;
  height: 12px;
  transform: rotate(90deg);
}
.write-a-post {
  cursor: pointer;
  font-size: 14px;
  padding-top: 3px;
}
.write-a-post:hover {
  color: #686868;
}
.search {
  cursor: pointer;
}
.search:hover {
  color: #686868;
}
.search img {
  width: 16px;
}
.logo {
  position: relative;
  left: 25px;
  height: 45px;
  width: 41px;
}
.logo img {
  height: 45px;
  cursor: pointer;
  width: 40px;
  object-fit: contain;
}
.sticky {
  position: fixed;
  top: 0;
}
@media only screen and (max-width: 768px) {
  .back-btn {
    display: none;
  }
  .write-a-post {
    display: none;
  }
}
@media only screen and (min-width: 280px) and (max-width: 576px) {
  .ham-menu {
    display: block;
    align-self: center;
  }
  .left-side {
    justify-content: center;
  }
  .logo {
    left: -10px;
  }
  .city-names {
    left: -6px;
  }
  #arrow {
    left: 0px;
  }
  .right-side {
    text-align: center;
  }
}
@media only screen and (min-width: 577px) and (max-width: 768px) {
  .ham-menu {
    display: block;
    align-self: center;
  }
  .left-side {
    justify-content: center;
  }
  .logo {
    left: -10px;
  }
  .city-names {
    left: -6px;
  }
  #arrow {
    left: 0px;
  }
  .right-side {
    text-align: center;
  }
}
</style>
