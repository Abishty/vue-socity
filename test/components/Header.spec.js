import { shallowMount, RouterLinkStub, createLocalVue } from '@vue/test-utils'
import Header from '@/components/Header'
import Dropdown from '@/components/Dropdown'

const FontAwesomeIconStub = '<i class="stub"></i>'

const transitionStub = () => ({
  render(h) {
    return this.$options._renderChildren
  }
})

const factory = () => {
  return shallowMount(Header, {
    mocks: {
      $route: {
        params: {
          city: 'delhi'
        }
      },
      data() {
        return {
          showDropdown: false,
          city: 'delhi'
        }
      }
    },
    stubs: {
      NuxtLink: RouterLinkStub,
      transition: transitionStub(),
      FontAwesomeIcon: FontAwesomeIconStub
    }
  })
}

describe('Header', () => {
  afterEach(() => {
    window.pageYOffset = 0
  })

  test('mounts properly', () => {
    const wrapper = factory()
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders properly', () => {
    const wrapper = factory()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Header renders with dropdown correctly', async () => {
    const wrapper = factory()
    wrapper.setData({ showDropdown: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('to check toggleDropdown function toggles the "showDropdown" variable', async () => {
    const wrapper = factory()
    wrapper.find('#arrow').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.showDropdown).toBe(true)
    wrapper.find('#arrow').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.showDropdown).toBe(false)
  })

  test('to see if Dropdown is rendered to the screen', async () => {
    const wrapper = factory()
    wrapper.setData({ showDropdown: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.find(Dropdown).exists()).toBeTruthy()
  })

  test('to check if clicking a city in the dropdown emits "cityChange" in the Header', async () => {
    const wrapper = factory()
    wrapper.setData({ showDropdown: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.find(Dropdown).exists()).toBeTruthy()
  })

  test('to check if the dropdown emit is reaching the Header', async () => {
    const wrapper = factory()
    wrapper.setData({ showDropdown: true })
    wrapper.setMethods({ changeCity: jest.fn() })
    await wrapper.vm.$nextTick()
    wrapper.find(Dropdown).vm.$emit('changeCity')
    expect(wrapper.vm.changeCity).toHaveBeenCalled()
  })

  test('to check if the handleScroll fn is adding and removing the sticky class on the header', async () => {
    const localVue = createLocalVue()
    const wrapper = shallowMount(Header, {
      localVue,
      attachToDocument: true,
      mocks: {
        $route: {
          params: {
            city: 'delhi'
          }
        }
      },
      stubs: {
        NuxtLink: RouterLinkStub,
        transition: transitionStub()
      }
    })
    const header = wrapper.find('.header-container')
    window.pageYOffset = 2000
    await wrapper.vm.handleScroll()
    expect(header.classes()).toContain('sticky')
    window.pageYOffset = 0
    await wrapper.vm.handleScroll()
    expect(header.classes()).not.toContain('sticky')
  })
})
