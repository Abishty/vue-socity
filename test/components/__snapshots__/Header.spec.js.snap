// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Header renders with dropdown correctly 1`] = `
<div>
  <div class="row header-container">
    <div class="col-lg-9 left-side">
      <div class="back-btn"><img src="../assets/arrow.svg"></div>
      <div class="logo"><a><img src="../assets/so_logo.svg"></a></div>
      <div class="city-names">Delhi</div>
      <div id="arrow"><img src="../assets/arrow.svg"></div>
    </div>
    <div class="col-lg-3 right-side">
      <div class="write-a-post">Write A Post</div>
      <div class="search"><img src="../assets/search.svg"></div>
    </div>
  </div>
  <dropdown-stub name="fade"></dropdown-stub>
</div>
`;

exports[`Header renders properly 1`] = `
<div>
  <div class="row header-container">
    <div class="col-lg-9 left-side">
      <div class="back-btn"><img src="../assets/arrow.svg"></div>
      <div class="logo"><a><img src="../assets/so_logo.svg"></a></div>
      <div class="city-names">Delhi</div>
      <div id="arrow"><img src="../assets/arrow.svg"></div>
    </div>
    <div class="col-lg-3 right-side">
      <div class="write-a-post">Write A Post</div>
      <div class="search"><img src="../assets/search.svg"></div>
    </div>
  </div>
  <!---->
</div>
`;
